File: app/src/main/java/ua/hospes/rtm/ui/race/TimerListController.java
package ua.hospes.rtm.ui.race;

import android.view.View;

import androidx.recyclerview.widget.RecyclerView;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import ua.hospes.rtm.core.ui.RxScrollListener;
import ua.hospes.rtm.widgets.TimeView;
import ua.hospes.rtm.utils.RxUtils;

/**
 * @author Andrew Khloponin
 */
class TimerListController extends RxScrollListener {
    private static final float AUTO_PLAY_AREA_START_PADDING_RELATIVE = 0.1f;
    private static final float AUTO_PLAY_AREA_END_PADDING_RELATIVE = 0.1f;

    private Set<TimeView> playingItems = new HashSet<>();
    private static final long SKIP_RECALCULATION_DURATION = 300;

    private long lastRecalculationTime;


    TimerListController() {
        RxUtils.manage(this, subject.debounce(SKIP_RECALCULATION_DURATION, TimeUnit.MILLISECONDS)
                .compose(RxUtils.applySchedulers())
                .subscribe(this::onActionCall));
    }

    private void onActionCall(RecyclerView recyclerView) {
        if (System.currentTimeMillis() < lastRecalculationTime + SKIP_RECALCULATION_DURATION) return;

        lastRecalculationTime = System.currentTimeMillis();

        playingItems = collectTimeViews(recyclerView);
    }

    @SuppressWarnings("ConstantConditions")
    private Set<TimeView> collectTimeViews(RecyclerView rv) {
        Set<TimeView> set = new HashSet<>();

        RecyclerView.LayoutManager lm = rv.getLayoutManager();

        int autoPlayAreaStart = (int) (rv.getTop() - rv.getHeight() * AUTO_PLAY_AREA_START_PADDING_RELATIVE);
        int autoPlayAreaEnd = (int) (rv.getBottom() + rv.getHeight() * AUTO_PLAY_AREA_END_PADDING_RELATIVE);

        int count = lm.getChildCount();
        for (int i = 0; i < count; i++) {
            View child = lm.getChildAt(i);
            int viewStart = lm.getDecoratedTop(child);
            int viewEnd = lm.getDecoratedBottom(child);

            boolean shouldPlay = false;
            shouldPlay = shouldPlay || (rv.getTop() <= viewStart && rv.getBottom() >= viewEnd); // completely visible
            shouldPlay = shouldPlay || !(autoPlayAreaStart > viewEnd || autoPlayAreaEnd < viewStart); // near center;

            if (shouldPlay) {
                RaceAdapter.MyHolder viewHolder = (RaceAdapter.MyHolder) rv.getChildViewHolder(child);
                set.add(viewHolder.sessionTimeView);
                set.add(viewHolder.driverTimeView);
            }
        }
        return set;
    }


    public void updateTime(long currentNanoTime) {
        for (TimeView tv : playingItems) tv.setCurrentNanoTime(currentNanoTime);
    }

    public void unsubscribe() {
        RxUtils.unsubscribe(this);
    }
}

File: app/src/main/java/ua/hospes/rtm/ui/race/TimerListController.kt
package ua.hospes.rtm.ui.race

import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.launch
import ua.hospes.rtm.core.ui.ReactiveScrollListener
import ua.hospes.rtm.widgets.TimeView
import java.util.*

private const val AUTO_PLAY_AREA_START_PADDING_RELATIVE = 0.1f
private const val AUTO_PLAY_AREA_END_PADDING_RELATIVE = 0.1f
private const val SKIP_RECALCULATION_DURATION: Long = 300

@Suppress("EXPERIMENTAL_API_USAGE")
internal class TimerListController(cs: CoroutineScope) : ReactiveScrollListener() {

    private var playingItems: Set<TimeView> = HashSet()

    private var lastRecalculationTime: Long = 0

    init {
        cs.launch { subject.asFlow().debounce(SKIP_RECALCULATION_DURATION).collect { onActionCall(it) } }
    }

    private fun onActionCall(recyclerView: RecyclerView) {
        if (System.currentTimeMillis() < lastRecalculationTime + SKIP_RECALCULATION_DURATION) return

        lastRecalculationTime = System.currentTimeMillis()

        playingItems = collectTimeViews(recyclerView)
    }

    private fun collectTimeViews(rv: RecyclerView): Set<TimeView> {
        val set = HashSet<TimeView>()

        val lm = rv.layoutManager

        val autoPlayAreaStart = (rv.top - rv.height * AUTO_PLAY_AREA_START_PADDING_RELATIVE).toInt()
        val autoPlayAreaEnd = (rv.bottom + rv.height * AUTO_PLAY_AREA_END_PADDING_RELATIVE).toInt()

        val count = lm!!.childCount
        for (i in 0 until count) {
            val child = lm.getChildAt(i)
            val viewStart = lm.getDecoratedTop(child!!)
            val viewEnd = lm.getDecoratedBottom(child)

            var shouldPlay = false
            shouldPlay = shouldPlay || rv.top <= viewStart && rv.bottom >= viewEnd // completely visible
            shouldPlay = shouldPlay || !(autoPlayAreaStart > viewEnd || autoPlayAreaEnd < viewStart) // near center;

            if (shouldPlay) {
                val viewHolder = rv.getChildViewHolder(child) as RaceAdapter.MyHolder
                set.add(viewHolder.sessionTimeView)
                set.add(viewHolder.driverTimeView)
            }
        }
        return set
    }


    fun updateTime(currentNanoTime: Long) {
        for (tv in playingItems) tv.currentNanoTime = currentNanoTime
    }
}